Rubberduck.CodeAnalysis/Inspections/Concrete/EmptyMethodInspection.cs

```diff
@@ -0,0 +1,131 @@
+﻿using Rubberduck.Inspections.Abstract;
+using Rubberduck.Inspections.Results;
+using Rubberduck.Parsing.Common;
+using Rubberduck.Parsing.Inspections.Abstract;
+using Rubberduck.Resources.Inspections;
+using Rubberduck.Parsing.VBA;
+using System.Collections.Generic;
+using System.Linq;
+using Rubberduck.Parsing.Symbols;
+using Rubberduck.Resources.Experimentals;
+using static Rubberduck.Parsing.Grammar.VBAParser;
+
+namespace Rubberduck.Inspections.Concrete
+{
+
+    /// <summary>
+    /// Identifies empty module member blocks.
+    /// </summary>
+    /// <why>
+    /// Methods containing no executable statements are misleading as they appear to be doing something which they actually don't.
+    /// This might be the result of delaying the actual implementation for a later stage of development, and then forgetting all about that.
+    /// </why>
+    /// <example hasResults="true">
+    /// <![CDATA[
+    /// Sub Foo()
+    ///     ' ...
+    /// End Sub
+    /// ]]>
+    /// </example>
+    /// <example hasResults="false">
+    /// <![CDATA[
+    /// Sub Foo()
+    ///     MsgBox "?"
+    /// End Sub
+    /// ]]>
+    /// </example>
+    [Experimental(nameof(ExperimentalNames.EmptyBlockInspections))]
+    internal class EmptyMethodInspection : InspectionBase// EmptyBlockInspectionBase
+    {
+        public EmptyMethodInspection(RubberduckParserState state)
+            : base(state) { }
+
+        protected override IEnumerable<IInspectionResult> DoGetInspectionResults()
+        {
+            // Exclude empty members in user interfaces, as long as all members of the interface are empty,
+            // since some VB users might use concrete user defined classes as interfaces,
+            // while RD marks them as interfaces all the same.
+
+            return UserDeclarations.OfType<ModuleBodyElementDeclaration>()
+                .Where(bodyElement => !BlockContainsExecutableStatements(bodyElement.Block))
+                .GroupBy(bodyElement => bodyElement.ComponentName)
+                // Exclude results from user interfaces
+                .Where(bodyElements => !State.DeclarationFinder.FindAllUserInterfaces()
+                                       // where all members of that interface contain no executables
+                                       .Where(interfaceModule => interfaceModule.ComponentName == bodyElements.Key
+                                                                && interfaceModule.Members.Count == bodyElements.Count())
```

Comment:
The `QualifiedModuleName` should be enough to know what we're looking at here; member count feels a bit arbitrary.

Reply:
But without count comparison we would be forced to inspect all members again, while in this way we can know for sure that the interface we are looking at is also used as concrete.

Reply:
I think we should care more about whether there is an `Implements` statement rather than matching the count of methods. Maybe look at the [Implement Interface refactoring](https://github.com/rubberduck-vba/Rubberduck/blob/189aa1ce552c0805e3591971c4389c494d02a92d/Rubberduck.Refactorings/ImplementInterface/ImplementInterfaceRefactoring.cs#L56) for an example of finding where interfaces are implemented. I think that if you can find a class that is `Implement`d by any other class, you can easily exclude it from the inspection, whether it has any body or not. (The question of whether an interface should have any code is a separate one anyway). If the class is not `Implement`'d by anyone else, then that is a candidate for the empty procedure inspection. Let the compiler take care of enforcing the member counts & parameter types.

Reply:
On the contrary, I'm not looking for a way to exclude all interfaces, but a way **to include some of them**.
I suspect that you guys think like C#'ers insread of VB's.
consider this test:

`[Test]
[Category("Inspections")]
public void EmptyMethodInConcreteClassMarkedAsInterface_ReturnsResult()
{
    const string interfaceCode =
        @"Public Sub DoSomething(ByVal a As Integer)
End Sub
Public Sub DoSomethingElse(ByVal a As Integer)
    MsgBox ""?""
End Sub";
    const string concreteCode =
        @"Implements IClass1

Private Sub IClass1_DoSomething(ByVal a As Integer)
    MsgBox ""?""
End Sub
Public Sub IClass1_DoSomethingElse(ByVal a As Integer)
    MsgBox ""?""
End Sub";`

This VB code is perfectly legit and and not that uncommon as RD users like ourselves might think... `interfaceCode` here (`IClass1`) represents not a real interface, but a concrete object class, which is only included in our `FindAllUserInterfaces()` because of the line `@"Implements IClass1` in the `concreteCode`.
This is one "interface" (or rather, !interface...) we want to include in this inspection, and the `Members.Count` compared to `bodyElements.Count()` (i.e, checking if there are more members then empty blocks detected for this `ClassModule`) and if true, including these results in our inspection, seems like the optimal solution at this time.

Sorry if I'm over complicating things... I'm one little duckling which is eager to learn

Reply:
Here's where I got confused. I thought the intention was to not consider the "interface" that contains code, like `IClass1` since even though it has some implementations, it is as you say, legal. But in this case, you are going to raise a result for the "interface" `IClass1`, rather than ignoring it. 

I'm a bit conflicted because I know some would use this more akin to an abstract class rather than an interface, and it can be done (albeit with some manual plumbing work), and in this case, it _shouldn't_ raise an inspection. On the other hand, if I'm writing a pure interface rather than an abstract class, I probably do want to get warned that I may have mucked up something. 

That said, more data is better than not enough, so I'm content to leave things as it is. :)

Reply:
Ah! I also got confused there, hadn't considered the complication of `@Interface` class modules that contained a concrete implementation for one or more members - a rather weird spot to be in IMO.
I think it could warrant its own separate "concrete implementation code in interface" inspection, to really drive home the idea that class modules whose default interface is intended to be implemented by other classes, should be abstract / not supply any implementation. If such "hybrid" interfaces are flagged, then it makes sense for *this inspection* to just ignore such modules.
On the other hand, there is currently no such inspection - a compromise could be to extract that part of the logic into its own `IsInterfaceWithConcreteMembers` function (or similar), so that it's 1) easier to tell *why* this condition is included (it *is* kind of an edge case), and 2) easier to pull out if/when a "concrete implementation in interface module" inspection is written.
Thoughts?

Reply:
> On the other hand, there is currently no such inspection - a compromise could be to extract that part of the logic into its own `IsInterfaceWithConcreteMembers` function (or similar), so that it's 1) easier to tell _why_ this condition is included (it _is_ kind of an edge case), and 2) easier to pull out if/when a "concrete implementation in interface module" inspection is written.
> Thoughts?

Makes sense. I think it will be easy to just go for the inspection idea, since we can use the `private bool BlockContainsExecutableStatements(BlockContext block)` in an abstract base for these two inspections.

Reply:
Inspections tend to have a rather shallow inheritance hierarchy though; this could be legit useful in other places beyond inspections too - a `ContainsExecutableStatements(this BlockContext block)` extension method seems appropriate.

Reply:
Actually - I'm pretty sure there's already something similar somewhere: the `EmptyXxxxBlock` inspections surely already do this... consider refactoring them to use the extension method if possible (otherwise it's technical debt we'll have to pay one day or another). @Hosch250 do you remember how it was done?

Reply:
It was done by `public class EmptyBlockInspectionListenerBase : VBAParserBaseListener, IInspectionListener` from which derived all others:
`EmptyCaseBlockInspection`, `EmptyDoWhileBlockInspection` and so on...
The entire functionality, however is encapsulated in this base class.

Reply:
Thanks for looking it up, I'm doing too many things at once! I really think the "does this block contain any executable statements" part should really exist in one single place. How about deriving this listener from the `EmptyBlockInspectionListenerBase`? That would solve the logic duplication problem I think.

Reply:
This was my approach in the first commit, but I re-implemented it as `InspectionBase` because this `InspectionListenerBase` is implementing a `QualifiedContextInspectionResults`, while @MDoerner suggested we should implement `DeclarationInspectionResult` instead.